build, libglusterfs: add --with-malloc=[tcmalloc,jemalloc,system] options

dmantipov opened the pull request:
build, libglusterfs: add --with-malloc option
    
Extend '--enable-tcmalloc' quirks to more generic
'--with-malloc=[tcmalloc,jemalloc,system]' option
to select Google's tcmalloc which is now the default,
FreeBSD's originated jemalloc or system malloc,
respectively. Also update glusterfs.spec.in so
tcmalloc is selected by default but '--with jemalloc'
and '--with system-malloc' is provided as well.
    
Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

mykaul commented:
I think we are somewhat challenged with supporting several memory allocation libraries, as we don't have a comprehensive test matrix for performance for all of them. Seems like tcmalloc without memory pools brings us some benefit, but when @mohit84 tried it with some patch I've proposed, he found it regressing. So I don't think there's any harm in adding more libraries, unsure of the real value (seem though any is better than glibc's... ?)

dmantipov commented:
> but when @mohit84 tried it with some patch I've proposed, he found it regressing

I can demonstrate 9-10% improvement with `tcmalloc` on simple `replica 3` volume with synthetic `fio` test.

mykaul commented:
> > but when @mohit84 tried it with some patch I've proposed, he found it regressing
> 
> I can demonstrate 9-10% improvement with `tcmalloc` on simple `replica 3` volume with synthetic `fio` test.

Yes, I've seen an improvement as well - but I don't know if we'll have the same / better / worse with jemalloc so hesitant to add more to the mix. (I'm assuming we'll move to disabling memory pools by default, so that removes one parameter from the matrix)

amarts reviewed (dismissed):
> so hesitant to add more to the mix. 

As it is just an option, and all our build scripts would pick the default one (or well tested one), its good to add it. It allows our users to test it quickly if needed.

@Shwetha-Acharya @Sheetalpamecha @deepshikhaaa Need to see if we have used `--enable-tcmalloc` anywhere in the build scripts.

amarts commented:
/run regression

deepshikhaaa commented:
> > so hesitant to add more to the mix.
> 
> As it is just an option, and all our build scripts would pick the default one (or well tested one), its good to add it. It allows our users to test it quickly if needed.
> 
> @Shwetha-Acharya @Sheetalpamecha @deepshikhaaa Need to see if we have used `--enable-tcmalloc` anywhere in the build scripts.

No, we have not used `--enable-tcmalloc` flag in the build scripts.

amarts commented:
@dmantipov I hope this doesn't have a conflict with #2675

dmantipov commented:
> @dmantipov I hope this doesn't have a conflict with #2675

Unfortunately it is. Since this is not too critical, let's do #2675 first, and then I adjust `--enable-malloc` stuff.

amarts commented:
Time to rebase this now.

dmantipov commented:
Done.

amarts commented:
/run regression

dmantipov commented:
Again, the whole thing is simple: let's make `tcmalloc` the default but take #2696 and corresponding `--with / --without` RPM spec options for the brave builders who understands what they're doing.

It's also desirable for `configure` to emit a warning if an allocator other than `tcmalloc` is enabled.

amarts commented:
Can we extend this PR to include changes for glusterfs.spec.in?

mohit84 commented:
> Can we extend this PR to include changes for glusterfs.spec.in?

Please merge this patch after merge the patch(https://github.com/gluster/glusterfs/pull/2773).

xhernandez reviewed (commented):
Shouldn't we use something like `--with-allocator <name>` instead of `--enable-xxx` ?

Based on autoconf documentation, `--enable` is more suited for optional features, while `--with` is more suited for package dependencies.

dmantipov commented:
Updated to select `tcmalloc` by default but provide `--with jemalloc` and `--with system-malloc` RPM spec options.

dmantipov commented:
Consistently use `-with-malloc=[tcmalloc,jemalloc,system]` in `configure.ac` and `--with tcmalloc` (default), `--with jemalloc` and `--with system-malloc` in `glusterfs.spec.in`.

dmantipov commented:
If `tcmalloc` is the default, should `configure` fail if it is not detected or issue a warning and fallback to system malloc? First approach seems breaks the most of current smoke tests because `tcmalloc` is missing within build environments.

amarts commented:
> If tcmalloc is the default, should configure fail if it is not detected or issue a warning and fallback to system malloc? First approach seems breaks the most of current smoke tests because tcmalloc is missing within build environments.

we have made tirpc as default, but without it also the configure continues to use system RPC. 

As we have configure output which emits the options choosen, I guess default ./configure should run without failure. Only when `--with tcmalloc` is provided, it should fail when tcmalloc dependencies are not met.

?

dmantipov commented:
> As we have configure output which emits the options choosen, I guess default ./configure should run without failure. Only when `--with tcmalloc` is provided, it should fail when tcmalloc dependencies are not met.

OK with this.

dmantipov commented:
Finally, how it is supposed to work:

1. For `configure`, `--with-malloc=[tcmalloc,jemalloc,system]` is provided. The first one is the default. If `tcmalloc` is not detected, then `configure` fails if `--with-malloc=tcmalloc` was specified, or issues the warning and switches to system malloc if no malloc implementation was specified. Also `configure` emits the warning if malloc implementation other than `tcmalloc` is used.
2. For `glusterfs.spec.in`, `--with tcmalloc`, which is the default (even if not specified), `--with jemalloc` and `--with system-malloc` options are provided.

xhernandez commented:
> > If tcmalloc is the default, should configure fail if it is not detected or issue a warning and fallback to system malloc? First approach seems breaks the most of current smoke tests because tcmalloc is missing within build environments.
> 
> we have made tirpc as default, but without it also the configure continues to use system RPC.
> 
> As we have configure output which emits the options choosen, I guess default ./configure should run without failure. Only when `--with tcmalloc` is provided, it should fail when tcmalloc dependencies are not met.
> 
> ?

Given that the regression is significant, I wouldn't fallback to system malloc, even if we show that tcmalloc is not enabled at the end of configure (most people doesn't check it, and automated builds won't check it for sure). This way we avoid that future releases have regressions in some distributions just because they forgot to install the libraries in the build system.

dmantipov commented:
> I wouldn't fallback to system malloc, even if we show that tcmalloc is not enabled at the end of configure 

This means that all environments currently used for automated builds (fedora, el7, centos7, freebsd) should have `tcmalloc` installed first.

xhernandez commented:
> > I wouldn't fallback to system malloc, even if we show that tcmalloc is not enabled at the end of configure
> 
> This means that all environments currently used for automated builds (fedora, el7, centos7, freebsd) should have `tcmalloc` installed first.

Yes. The alternative is that all initial Gluster 10 releases will have a significant performance regression on those distributions. I think it's better to cause a failure in the build process.

amarts reviewed (dismissed):
LGTM

Commit:
build, libglusterfs: add --with-malloc option

Extend '--enable-tcmalloc' quirks to more generic
'--with-malloc=[tcmalloc,jemalloc,system]' option
to select Google's tcmalloc which is now the default,
FreeBSD's originated jemalloc or system malloc,
respectively. Also update glusterfs.spec.in so
tcmalloc is selected by default but '--with jemalloc'
and '--with system-malloc' is provided as well.

Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

dmantipov commented:
/recheck smoke

rakshithakamath94 commented:
/recheck smoke

amarts commented:
/run regression

amarts commented:
@mohit84 @xhernandez see if this can be taken in now.

xhernandez commented on configure.ac:
```diff
@@ -337,22 +337,65 @@ if test "x$enable_ubsan" = "xyes"; then
+        if test x$with_malloc = xtcmalloc; then
+            AC_MSG_ERROR([unable to detect tcmalloc, make sure it's installed.])
+        else
+            AC_MSG_WARN([unable to detect tcmalloc, using system malloc instead.])
```
I think that we should enforce tcmalloc presence always. If the user doesn't want to install it, an explicit `--with-malloc=system` should be provided to avoid the error.

dmantipov replied:
OK.

amarts reviewed (approved):
Can this be tested on a EL8 system before merging? so that there is no issues?

amarts commented:
/run regression

mohit84 commented:
> Can this be tested on a EL8 system before merging? so that there is no issues?

I don't this it has anything specific to EL8  will try to test later, current the systems are busy in some other perf test case.

gluster-ant commented:
1 test(s) failed 
./tests/basic/afr/granular-esh/granular-indices-but-non-granular-heal.t

0 test(s) generated core 


3 test(s) needed retry 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/basic/afr/granular-esh/granular-indices-but-non-granular-heal.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/1664/

dmantipov commented:
/run full regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/glusterfs/bug-866459.t

0 test(s) generated core 


2 test(s) needed retry 
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/bugs/glusterfs/bug-866459.t
